sound/soc/sof/ipc.c

```diff
@@ -671,23 +772,67 @@ int snd_sof_ipc_set_get_comp_data(struct snd_sof_ipc *ipc,
 	cdata->rhdr.hdr.cmd = SOF_IPC_GLB_COMP_MSG | ipc_cmd;
 	cdata->cmd = ctrl_cmd;
 	cdata->type = ctrl_type;
-	cdata->rhdr.hdr.size = scontrol->size;
 	cdata->comp_id = scontrol->comp_id;
-	cdata->num_elems = scontrol->num_channels;
+	cdata->msg_id = 0;
+
+	/* calculate header and data size */
+	switch (cdata->type) {
+	case SOF_CTRL_TYPE_VALUE_CHAN_GET:
+	case SOF_CTRL_TYPE_VALUE_CHAN_SET:
+		sparams.msg_bytes = scontrol->num_channels *
+			sizeof(struct sof_ipc_ctrl_value_chan);
+		sparams.hdr_bytes = sizeof(struct sof_ipc_ctrl_data);
+		sparams.elems = scontrol->num_channels;
+		break;
+	case SOF_CTRL_TYPE_VALUE_COMP_GET:
+	case SOF_CTRL_TYPE_VALUE_COMP_SET:
+		sparams.msg_bytes = scontrol->num_channels *
+			sizeof(struct sof_ipc_ctrl_value_comp);
+		sparams.hdr_bytes = sizeof(struct sof_ipc_ctrl_data);
+		sparams.elems = scontrol->num_channels;
+		break;
+	case SOF_CTRL_TYPE_DATA_GET:
+	case SOF_CTRL_TYPE_DATA_SET:
+		sparams.msg_bytes = cdata->data->size;
+		sparams.hdr_bytes = sizeof(struct sof_ipc_ctrl_data) +
+			sizeof(struct sof_abi_hdr);
+		sparams.elems = cdata->data->size;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	cdata->rhdr.hdr.size = sparams.msg_bytes + sparams.hdr_bytes;
+	cdata->num_elems = sparams.elems;
+	cdata->total_elems = sparams.elems;
```

Comment:
I don't get that part. Above you described 
````c
uint32_t num_elems;	/**< in array elems or bytes for data type */
 uint32_t total_elems;   /**< total size if message is sent in parts */
````

so why do you need the two fields if they contain the same value? Looks like the num_elems will be overwritten anyways if you send multiple packets, right?

Reply:
I'm just filling the member needed for large messages with the same value as "small" messages. Here we are still in "small" message processing. It could be left out if it is confusing...

Reply:
So now understanding what you are asking, so at least total_elems can be used in fw side to reserve enough memory for new parameters and also possibly check that all messages have been received.

Reply:
You may want to add a comment that for BINARY data sent in one shot, num_elems and total_elems are identical, while for multiple packets num_elems corresponds to the size of the current payload (which will typically be different for the last packet) and total_elems remains fixed as the total size.